src/lem/eval.rs

```diff
@@ -9,63 +9,880 @@ pub(crate) fn eval_step() -> Func {
     let apply_cont = apply_cont();
     let make_thunk = make_thunk();
 
-    func!((expr_in, env_in, cont_in): 3 => {
-        let (expr, env, cont, ctrl) = reduce(expr_in, env_in, cont_in);
+    func!((expr, env, cont): 3 => {
+        let (expr, env, cont, ctrl) = reduce(expr, env, cont);
         let (expr, env, cont, ctrl) = apply_cont(expr, env, cont, ctrl);
-        let (expr, env, cont, ctrl) = make_thunk(expr, env, cont, ctrl);
+        let (expr, env, cont, _ctrl) = make_thunk(expr, env, cont, ctrl);
         return (expr, env, cont)
     })
-    .unwrap()
+}
+
+fn safe_uncons() -> Func {
+    func!((xs): 2 => {
+        let nil: Expr::Nil;
```

Comment:
Idea: receive a `nil` as an argument to avoid creating multiple nil pointers

Reply:
`nil` pointers are created only once, even if used on different `func`s because of the global allocation and because `func`s  are inlined (not treated as a blackbox) inside other funcs